### PR TITLE
Refactor code, move condition checks into a for-loop

### PR DESCRIPTION
**Pull request by hainesc**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Refactor the code, move all condition checks into a for-loop

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

None
```release-note

```

### CODE REVIEW

**Comment by hainesc** [comment]
Maybe it is better way to check so many conditions. Inspired by [dexidp/dex](https://github.com/dexidp/dex), See the code https://github.com/dexidp/dex/blob/master/cmd/dex/serve.go#L76

**Comment by hainesc** [comment]
CLA signed

**Review by neolit123 — COMMENTED** [review]
thanks for the PR
/release-note-none
/ok-to-test
/kind cleanup
/priority backlog

**Comment by hainesc** [comment]
@neolit123 Could you review the code again? thanks.

**Review by neolit123 — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go [hunk 1]
```diff
@@ -70,24 +70,21 @@ const (
...
+	checks := []struct {
+		bad    bool
+		errMsg string
+	}{
+		{lec.LeaseDuration <= lec.RenewDeadline, "leaseDuration must be greater then renewDeadline"},
+		{lec.RenewDeadline <= time.Duration(JitterFactor*float64(lec.RetryPeriod)), "renewDeadline must be greater than retryPeriod*JitterFactor"},
+		{lec.LeaseDuration < 1, "leaseDuration must be greater than zero"},
+		{lec.RenewDeadline < 1, "renewDeadline must be greater than zero"},
+		{lec.RetryPeriod < 1, "retryPeriod must be greater than zero"},
+		{lec.Lock == nil, "Lock must not be nil."},
```
Comment (cheftako) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Should be "lock" not "Lock"

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go [hunk 2]
```diff
@@ -70,24 +70,21 @@ const (
...
-	}
-	if lec.RetryPeriod < 1 {
-		return nil, fmt.Errorf("retryPeriod must be greater than zero")
-	}
-
-	if lec.Lock == nil {
-		return nil, fmt.Errorf("Lock must not be nil.")
+	for _, check := range checks {
+		if check.bad {
+			return nil, fmt.Errorf(check.errMsg)
```
Comment (cheftako) [inline staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: I dislike having the Errorf here. We do not *currently* have any formatted output fields in these error messages, but this precludes us from doing so. I think we need to retain that ability. You could accomplish that by changing the struct contain an error and doing the fmt.Errorf earlier. However we are then doing more work than in the current manner.

**Review by cheftako — COMMENTED** [review]
One minor lint issue. However a little worried about losing the ability to format values into the error messages.

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go (hunk 2 above)
Comment (hainesc) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Yes, as you shown, it precludes us from making a formatted `error` message here. It is a better choice only when all `error` messages can be constructed from a string, for formatted ones, eg, `fmt.Error("A should be greater than %d", threshold)`, we can just add code before or below this piece of code.
`if A < threshold {
    return nil, fmt.Errorf("A should be greater than %d", threshold)
}`

Comment (hainesc) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: > I dislike having the Errorf here. We do not _currently_ have any formatted output fields in these error messages, but this precludes us from doing so. I think we need to retain that ability. You could accomplish that by changing the struct contain an error and doing the fmt.Errorf earlier. However we are then doing more work than in the current manner.

Yes, as you shown, it precludes us from making a formatted `error` message here. It is a better choice only when all `error` messages can be constructed from a string, for formatted ones, eg, `fmt.Error("A should be greater than %d", threshold)`, we can just add code before or below this piece of code.
`if A < threshold { return nil, fmt.Errorf("A should be greater than %d", threshold) }`

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go (hunk 1 above)
Comment (hainesc) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: > Should be "lock" not "Lock" (I realize this is a pre-existing problem)

Improved, thanks

File: staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go (hunk 2 above)
Comment (cheftako) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: Was thinking about this and if you really care you can always call  fmt.sprintf when constructing the message.
Eg.
{lec.LeaseDuration <= lec.RenewDeadline, fmt.Sprintf("leaseDuration(%d) must be greater than renewDeadline(%d)", lec.LeaseDuration, lec.RenewDeadline)},

Comment (hainesc) [reply staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go]: > Was thinking about this and if you really care you can always call fmt.sprintf when constructing the message.
> Eg.
> {lec.LeaseDuration <= lec.RenewDeadline, fmt.Sprintf("leaseDuration(%d) must be greater than renewDeadline(%d)", lec.LeaseDuration, lec.RenewDeadline)},

Good idea.

**Comment by hainesc** [comment]
ping @mikedanese @timothysc 
Hi there, would you please have a consideration to label it as `/approved`?

**Comment by mikedanese** [comment]
We don't do validation like this anywhere else. I don't see the benefit of making this inconsistent and it doesn't make the code clearer in my opinion. Table driven tests are useful when the assertions in the loop are more complex than `if bad {}`.

**Pull request closed without merging**